packages/mui-base/src/MenuUnstyled/MenuButtonUnstyled.tsx

```diff
@@ -0,0 +1,105 @@
+import * as React from 'react';
+import {
+  // unstable_useForkRef as useForkRef,
+  unstable_useControlled as useControlled,
+} from '@mui/utils';
+import { useSlotProps } from '../utils';
+
+export interface MenuButtonProps {
+  className?: string;
+  label?: string;
+  open?: boolean;
+  onOpenChange?: (
+    event: React.MouseEvent<HTMLButtonElement> | React.KeyboardEvent<HTMLButtonElement>,
+    open: boolean,
+  ) => void;
+  defaultOpen?: boolean;
+  children?: React.ReactNode;
+  slots?: {
+    root?: React.ElementType;
+  };
+  slotProps?: {
+    root?: React.HTMLAttributes<HTMLButtonElement>;
+  };
+}
+
+interface MenuButtonContextState {
+  open: boolean;
+}
+
+export const MenuButtonContext = React.createContext<MenuButtonContextState | null>(null);
+
+const MenuButton = React.forwardRef(function MenuButton(
+  props: MenuButtonProps,
+  ref: React.ForwardedRef<HTMLButtonElement>,
+) {
+  const {
+    children,
+    label,
+    open: openProp,
+    defaultOpen,
+    onOpenChange,
+    slots = {},
+    slotProps = {},
+    ...other
+  } = props;
+  const focusableElement = React.useRef<HTMLElement | null>(null);
+
+  const [isOpen, setOpen] = useControlled({
+    controlled: openProp,
+    default: defaultOpen,
+    name: 'MenuButton',
+  });
+
+  const contextValue = React.useMemo(() => ({ open: isOpen }), [isOpen]);
+
+  const handleClick = React.useCallback(
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      setOpen((open) => {
+        return !open;
+      });
+
+      onOpenChange?.(event, !isOpen);
+    },
+    [isOpen, setOpen, onOpenChange],
+  );
+
+  const handleKeyDown = React.useCallback(
+    (event: React.KeyboardEvent<HTMLButtonElement>) => {
+      if (event.key === 'ArrowDown') {
+        event.preventDefault();
+        setOpen(true);
+        onOpenChange?.(event, true);
+      }
+    },
+    [setOpen, onOpenChange],
+  );
+
+  React.useEffect(() => {
+    if (isOpen) {
+      focusableElement.current?.focus();
+    }
+  }, [isOpen]);
+
+  const Root = slots.root || 'button';
+  const rootProps = useSlotProps({
+    elementType: Root,
+    externalForwardedProps: other,
+    externalSlotProps: slotProps.root,
+    additionalProps: {
+      ref,
+      onClick: handleClick,
+      onKeyDown: handleKeyDown,
+    },
+    ownerState: { ...props, open: isOpen },
+  });
+
+  return (
+    <MenuButtonContext.Provider value={contextValue}>
+      <Root {...rootProps}>{label}</Root>
+      {children}
```

Comment:
Shouldn't the MenuButton pass these props?
```suggestion
      {React.cloneElement(children, { anchorEl, ref })}
```

One important reason of having a MenuButton component (in my opinion) is that developers don't have to declare the `anchorEl` and `open` at the top of the component. The usage should be like this:

```js
<MenuButtonUnstyled
  slots={{ root: MenuButton }}
  label="My account"
>
  <MenuUnstyled
    slots={{ root: Popper, listbox: StyledListbox }}
    slotProps={{ root: { placement: 'bottom-start' }, listbox: { id: 'simple-menu' } }}
  >
```

Reply:
cc @oliviertassinari

Reply:
Having `{React.cloneElement(children, { anchorEl, ref })}` prevents developers from having intermediate components between MenuButton and Menu (or a Fragment). I'll try to achieve this with a context instead.

Reply:
+1 for using context